File: acm/3nPlus1.py
class Solver3nPlus1:
    def is_even(self, n):
        return n % 2 == 0

    def maxCycle(self, n):
        cycle = 1
        while n != 1:
            if self.is_even(n):
                n = n / 2
            else:
                n = (3 * n) + 1
            cycle += 1
        return cycle

    def maxCycleRange(self, i, j):
        maxCycle = 0
        for x in range(i, j + 1):
            maxCycle = max(maxCycle, self.maxCycle(x))
        return maxCycle

if __name__ == "__main__":
    s = Solver3nPlus1()
    print s.maxCycleRange(1, 10)
    print s.maxCycleRange(100, 200)
    print s.maxCycleRange(201, 210)
    print s.maxCycleRange(900, 1000)

File: acm/3nPlus1.go
package main

import "fmt"

func is_even(n int) (eveness bool) {
	return n%2 == 0
}

func max(a int, b int) (result int) {
	if a > b {
		return a
	}
	return b
}

func Solver3nPlus1(i int, j int) (max_cycles int) {
	result := 0
	for x := i; x < j; x++ {
		n := x
		current_cycle := 0
		for n != 1 {
			if is_even(n) {
				n = n / 2
			} else {
				n = (3 * n) + 1
			}
			current_cycle++
		}
		result = max(result, current_cycle)
	}
	return result + 1
}
func main() {
	fmt.Println(Solver3nPlus1(1, 10))
	fmt.Println(Solver3nPlus1(100, 200))
	fmt.Println(Solver3nPlus1(201, 210))
	fmt.Println(Solver3nPlus1(900, 1000))
}
